### PR TITLE
start consensus only after syncing with the network

### PR DESCRIPTION
**Pull request by lorenzo-dev1**
This PR aims at providing a quick mitigation for one slashing issue that we saw on Piccadilly. More specifically the issue happens when we start the consensus engine before syncing to the current chain head. This can cause equivocation to be done by the node, because he is going to vote for an old height in which he might have already voted.

Please note that this is not a final fix, it just aims at preventing the most trivial edge case. The question of figuring out whether we are synced or not is not trivial at all, and requires more work to be implemented properly.

Moreover this solution will not be effective if the validator is under attack by malicious node who wants to impede its syncing. This specific problem requires modification on the sync mechanism, exploiting the existance of a segregated consensus network (https://github.com/autonity/autonity/pull/923)

The consensus engine now will start only if:
- we are already in sync with the network
- we completed a chain sync with a peer
- we forced mining start via:
  - devmode
  - call to miner api
  - using the `--mine` flag

### CODE REVIEW

File: miner/miner.go [hunk 1]
```diff
@@ -112,34 +134,42 @@ func (miner *Miner) update() {
...
-				if shouldStart {
-					miner.worker.start()
-				}
-			case downloader.DoneEvent:
-				canStart = true
-				if shouldStart {
-					miner.worker.start()
+				miner.syncFailures++
+				miner.eth.Logger().Info("Chain syncing failed", "#failures", miner.syncFailures, "shouldStart", miner.shouldStart, "canStart", miner.canStart)
+				// TODO(lorenzo) discuss
```
Comment (lorenzo-dev1) [inline miner/miner.go:141]: Please if you have any feedback on this approach let me know, I am not 100% sure this is the best way to go about it.

I am thinking about scenarios like:
- epoch period 1800
- we are at height=4000
- our node $x$ was offline and he is at height = 100. He was in the committee at that time. He is still in committee at height 4000 (but he doesn't know it yet).
- node $x$ comes back online, but he is attacked by malicious nodes which advertise height >> 4000 as their chain head
- node $x$ will keep trying to sync with the malicious nodes (since they claim to have the higher total difficulty chain). Once a sync fails, the malicious peer gets dropped. However the attacker keep spinning up fresh nodes (they don't have to be validators) and connecting them to $x$, thus keeping the attack going.
- node $x$ will stay stuck at height=100, without possibility of syncing.

After several failures, we have two possibilities:
- we do not start consensus, the attacker succeeds. He can stop our consensus until he manages to keep connecting nodes to $x$
- we start consensus anyways, hoping that we are synced.
  - if we are synced, all good  
  -  If we are not like in the example, we will start asking sync to other nodes, which will reply with consensus messages for height = 4000. However we cannot even verify those (we do not know the committee).

Block and block hashes announcements will not help, since they will be discarded because too far (blockwise). see:
- https://github.com/autonity/autonity/blob/develop/eth/fetcher/block_fetcher.go#L399 for hashes
- https://github.com/autonity/autonity/blob/develop/eth/fetcher/block_fetcher.go#L757 for blocks

So basically they will stop us from syncing. We can start consensus and hope we are synced (risking equivocating until we have a WAL) or just do not start consensus.

Comment (lorenzo-dev1) [reply miner/miner.go:141]: I am wondering if the best solution if we keep failing to sync would be to crash the client and log a message to the operator explaining what is likely happening. Then the operator can connect only to a set of trusted peers to sync and restart consensus.

Another option would be to feed additional information from:
- consensus (askSync)
- announcement

To understand if we are actually synced or not.

Comment (lorenzo-dev1) [reply miner/miner.go:141]: Originally posted by @tbssajal here https://github.com/autonity/autonity/pull/924#discussion_r1450409852
 
> - epoch period 1800
> - we are at height=4000
> - our node $x$ was offline and he is at height = 100. He was in the committee at that time. He is still in committee at height 4000 (but he doesn't know it yet).

It doesn't sound right. So the epochs are 0 -> for blocks [0, 1800), 1 -> for blocks [1800, 3600), 2 -> for blocks [3600, 5400). If $x$ is at height 100 and we are at height 4000, that means $x$ was inactive the whole epoch 1 -> for blocks [1800, 3600). So $x$ should not be selected as a committee member for epoch 2 by the protocol as he was inactive in in the previous epoch and if some sort of mechanism punishes inactivity then $x$ would probably be in jail for epoch 2.

Comment (lorenzo-dev1) [reply miner/miner.go:141]: At the moment inactive validators are not punished, so this is possible. Until we have the omission failure in place, I don't think we can rule out this scenario.

Comment (lorenzo-dev1) [reply miner/miner.go:141]: Originally posted by @tbssajal here https://github.com/autonity/autonity/pull/924#discussion_r1450575225

At the moment, yes we cannot. But I was thinking there should be a simple mechanism that does not punishes or rewards anyone but takes some sort of attendance before selecting committee members. For example right now validators need to be in `Active` state to be in committee. There can be a simple mechanism via contract that will change validator state as a form of attendance and only validators who submit attendance in the previous epoch can be considered as a candidate for the committee of the next epoch. There should be a separate issue for this I suppose.

Comment (tbssajal) [reply miner/miner.go:141]: > There can be a simple mechanism via contract that will change validator state as a form of attendance and only validators who submit attendance in the previous epoch can be considered as a candidate for the committee of the next epoch.

I created a discussion in response: https://github.com/clearmatics/autonity-protocol/discussions/269

Comment (lorenzo-dev1) [reply miner/miner.go:141]: I replied to your proposal in the protocol repo. I will mark it this discussion as solved for the moment, since this PR scope has been redefined and now it only tries to provide a quick fix for cases where there is no attack going on.

File: miner/miner.go [hunk 2]
```diff
@@ -68,40 +71,56 @@ type Miner struct {
 	stopCh  chan struct{}
 
 	wg sync.WaitGroup
+
+	// used in the miner update loop
+	syncFailures uint64 // counts how many times chain sync failed with remote peers
+	canStart     bool
```
Comment (lorenzo-dev1) [inline miner/miner.go:77]: Still need to figure out whether `canStart` and `events` should be protected for concurrent access. They are both accessed by the `update()` loop and by the `ForceStartMining` method (which is called by the API or when starting the client).

I think that theoretically there is a concurrent access possibility, but practically it should not be a problem. Still thinking about it though

File: eth/sync.go [hunk 3]
```diff
@@ -62,11 +62,12 @@ func (h *handler) syncTransactions(p *eth.Peer) {
...
-	force       *time.Timer
-	forced      bool // true when force timer fired
-	peerEventCh chan struct{}
-	doneCh      chan error // non-nil when sync is running
+	handler         *handler
+	force           *time.Timer
+	forced          bool // true when force timer fired
+	peerEventCh     chan struct{}
+	doneCh          chan error // non-nil when sync is running
+	syncedEventSent bool       // true when the downloader.SyncedEvent has been sent (we are synced with our current peerset)
```
Comment (Jason-Zhangxin-Chen) [inline eth/sync.go:70]: How is the lifecycle of the chainSyncer object managed? Is it created for each sync procedure? In this case, the flag should work.

File: miner/miner.go (hunk 2 above)
Comment (Jason-Zhangxin-Chen) [reply miner/miner.go:77]: I think we need to use RWMutex, if the two routine are just readers of them, just apply readLock while the writter should apply writeLock.

File: miner/miner.go [hunk 4]
```diff
@@ -68,40 +71,56 @@ type Miner struct {
...
+	// it will be true when node is a committee member, therefore consensus engine should be started if possible.
+
+	/* miner.canStart is set at initialization to false
+	* miner.canStart = true when consensus engine can be safely started OR if we forced mining start.
+	* It is safe to start the consensus engine when we are reasonably sure to be synced with the head of the chain:
+	* - the first chain sync with our peers terminates, and we conclude that we are synced to the chain head.
+	* - the first chain sync with our peer terminates, and we successfully imported blocks till the head of the chain
+	*
+	* NOTE: This mechanism does not give 100% guarantee that we are synced to the head of the chain before starting consensus. There is always the possibility that we are connected to peers which are behind the "global" chain head.
+	 */
```
Comment (Jason-Zhangxin-Chen) [inline miner/miner.go:122]: Is it possilble for a connected malicious peer who announces a block been mined in the future height can cause the stop of the mining in local node?

File: miner/miner.go [hunk 5]
```diff
@@ -112,34 +131,39 @@ func (miner *Miner) update() {
...
+				miner.eth.Logger().Info("Chain syncing failed", "#failures", miner.syncFailures, "shouldStart", miner.shouldStart, "canStart", miner.canStart)
+				// if we fail more than maxSyncFailures times consequently, assume we are under attack
+				if miner.syncFailures >= maxSyncFailures {
+					miner.eth.Logger().Warn("************************** SYNC ATTACK DETECTED **************************")
+					miner.eth.Logger().Warn("Multiple sequential chain sync failures detected", "sync failures", miner.syncFailures)
+					miner.eth.Logger().Warn("Your node is probably under attack by malicious peers, which are preventing your node from syncing")
+					miner.eth.Logger().Warn("Try restarting your node and connecting to a trusted set of peers")
+					miner.eth.Logger().Warn("Reach out to Autonity social media channels for support and additional informations")
+					miner.eth.Logger().Warn("**************************************************************************")
+					panic("sync attack detected")
```
Comment (Jason-Zhangxin-Chen) [inline miner/miner.go:146]: Do we really panic when we are under sync attack?

File: eth/sync.go (hunk 3 above)
Comment (lorenzo-dev1) [reply eth/sync.go:70]: There is only one chainSyncer that is created at node initialization. The syncedEventSent is not accessed concurrently however, only a single go routine ever reads/writes it.

File: miner/miner.go (hunk 4 above)
Comment (lorenzo-dev1) [reply miner/miner.go:122]: Yeah that is possible, because we always try to sync with the peer which advertises the highest height. However once we managed to sync successfully once we will stop reacting to downloader events.

File: miner/miner.go (hunk 5 above)
Comment (lorenzo-dev1) [reply miner/miner.go:146]: Yes, I think it is the safest way to handle this scenario currently. This is because if we are under a sync attack (nodes advertising height >> actual current height, and then failing sync) we will not be able to actually sync the chain. Therefore we cannot start consensus anyways.

The proper solution will require modifications in the syncing mechanism (e.g. relying on 2/3 honesty assumption of committee) but it is too complicated to be included into this release. This fix just aims at avoiding node getting slashed because of stop-start.

At the moment I feel like the only way to mitigate the issue is to crash the node (since we cannot start consensus anyways) and notify the operator, which can then connect to some trusted peers to actually make its node sync and start consensus.

If any better suggestions let me know.

Comment (lorenzo-dev1) [reply miner/miner.go:146]: Please note that we just need to sync successfully once and this attack will not work anymore. The attacker needs to basically eclipse us at startup for this to work.

File: miner/miner.go (hunk 2 above)
Comment (lorenzo-dev1) [reply miner/miner.go:78]: Solved by https://github.com/autonity/autonity/pull/927/commits/d2b3c521c61561c79e41228a412649a9d11f5f8a via code reorganization. Now concurrent access to canStart and events cannot happen anymore.

**Review by yazzaoui — APPROVED** [review]
LGTM !

**Review by vpiyush — APPROVED** [review]
LGTM
